core/src/avm2/class.rs

```diff
@@ -28,6 +32,82 @@ bitflags! {
     }
 }
 
+/// A function that can be used to allocate instances of a class.
+///
+/// By default, the `implicit_deriver` is used, which attempts to use the base
+/// class's deriver, and defaults to `ScriptObject` otherwise. Custom derivers
+/// anywhere in the class inheritance chain can change the representation of
+/// all subtypes that use the implicit deriver.
+///
+/// Parameters for the deriver are:
+///
+///  * `constr` - The class constructor that was called (or will be called) to
+///  construct this object. This must be the current class (using a base class
+///  will cause the wrong class to be read for traits).
+///  * `proto` - The prototype attached to the class constructor.
+///  * `activation` - This is the current AVM2 activation.
+pub type DeriverFn = for<'gc> fn(
+    Object<'gc>,
+    Object<'gc>,
+    &mut Activation<'_, 'gc, '_>,
+) -> Result<Object<'gc>, Error>;
+
+#[derive(Clone, Collect)]
+#[collect(require_static)]
+pub struct Deriver(pub DeriverFn);
+
+impl fmt::Debug for Deriver {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_tuple("Deriver")
+            .field(&"<native code>".to_string())
+            .finish()
+    }
+}
+
+/// The implicit deriver for new classes.
+///
+/// This attempts to use the parent type's deriver, and if such a deriver does
+/// not exist, we default to `ScriptObject`.
+pub fn implicit_deriver<'gc>(
+    mut constr: Object<'gc>,
+    proto: Object<'gc>,
+    activation: &mut Activation<'_, 'gc, '_>,
+) -> Result<Object<'gc>, Error> {
+    let mut base_constr = Some(constr);
+    let mut base_class = constr.as_class();
+    let mut instance_deriver = None;
+
+    while let (Some(b_constr), Some(b_class)) = (base_constr, base_class) {
+        let base_deriver = b_class.read().instance_deriver();
+
+        if base_deriver as usize != implicit_deriver as usize {
+            instance_deriver = Some(base_deriver);
+            break;
+        }
+
+        base_constr = b_constr.base_class_constr();
+        base_class = base_constr.and_then(|c| c.as_class());
+    }
```

Comment:
Instead of doing this at object construction time, I wonder if this be done once when a class is created, and cache the specific deriver for later.

Reply:
It's a little too early to do it in `Class`, since that only holds static information about a class (e.g. names, traits, and so on). At the point that `Class` gets constructed all it gets is the multiname to look up it's superclass with.

However, caching this in `ClassObject` would work.